File: src/bin/union.cpp
#include <inttypes.h>
#include <iostream>
#include <tuple>
#include <variant>

enum Protocol { UDP,
                TCP };
struct IP {
    uint32_t address;
    static IP from_u32(uint32_t address) {
        IP ip;
        ip.address = address;
        return ip;
    }
    uint32_t to_u32() { return this->address; }
    static IP from_u8(uint8_t a, uint8_t b, uint8_t c, uint8_t d) {
        IP ip;
        ip.address = (a << 24) | (b << 16) | (c << 8) | d;
        return ip;
    }
    std::tuple<uint8_t, uint8_t, uint8_t, uint8_t> to_u8() {
        uint8_t a = (this->address >> 24) & 0xFF;
        uint8_t b = (this->address >> 16) & 0xFF;
        uint8_t c = (this->address >> 8) & 0xFF;
        uint8_t d = this->address & 0xFF;
        return std::make_tuple(a, b, c, d);
    }
    std::string to_string() {
        auto [a, b, c, d] = this->to_u8();
        return std::to_string(a) + "." + std::to_string(b) + "." + std::to_string(c) + "." + std::to_string(d);
    }
};
struct FiveTuple {
    IP source_ip;
    IP destination_ip;
    uint16_t source_port;
    uint16_t destination_port;
    Protocol protocol;
};
using SourceData = std::variant<IP, FiveTuple>;
void print_data(SourceData data) {
    if (std::holds_alternative<IP>(data)) {
        IP ip = std::get<IP>(data);
        std::cout << "IP: " << ip.to_string() << std::endl;
    } else if (std::holds_alternative<FiveTuple>(data)) {
        FiveTuple ft = std::get<FiveTuple>(data);
        std::cout << "FiveTuple: (" << ft.source_ip.to_string() << ", " << ft.destination_ip.to_string() << ", " << ft.source_port << ", " << ft.destination_port << ", " << ft.protocol << ")" << std::endl;
    }
}
int main() {
    print_data(IP::from_u8(192, 168, 0, 1));
    print_data(FiveTuple{
        source_ip : IP::from_u8(192, 168, 0, 1),
        destination_ip : IP::from_u8(192, 168, 0, 2),
        source_port : 55555,
        destination_port : 53,
        protocol : Protocol::UDP,
    });
    print_data(FiveTuple{
        source_ip : IP::from_u8(127, 0, 0, 1),
        destination_ip : IP::from_u8(127, 0, 0, 1),
        source_port : 54321,
        destination_port : 8080,
        protocol : Protocol::TCP,
    });
    return 0;
}

File: src/bin/union.rs
struct IP(u32);
impl IP {
    fn to_string(self) -> String {
        let ip = self.0;
        format!(
            "{}.{}.{}.{}",
            (ip >> 24) as u8,
            (ip >> 16) as u8,
            (ip >> 8) as u8,
            ip as u8
        )
    }
}
impl From<(u8, u8, u8, u8)> for IP {
    fn from(ip: (u8, u8, u8, u8)) -> IP {
        IP((ip.0 as u32) << 24 | (ip.1 as u32) << 16 | (ip.2 as u32) << 8 | ip.3 as u32)
    }
}
impl From<&str> for IP {
    fn from(ip: &str) -> IP {
        let ip: Vec<u8> = ip
            .split('.')
            .map(|x| x.parse::<u8>().unwrap())
            .collect::<Vec<u8>>();
        (ip[0], ip[1], ip[2], ip[3]).into()
    }
}
#[derive(Debug)]
enum Protocol {
    TCP,
    UDP,
}
struct FiveTuple {
    source_ip: IP,
    destination_ip: IP,
    source_port: u16,
    destination_port: u16,
    protocol: Protocol,
}
enum SourceData {
    IP(IP),
    FiveTuple(FiveTuple),
}
fn print_data(source_data: SourceData) {
    match source_data {
        SourceData::IP(ip) => println!("IP: {}", ip.to_string()),
        SourceData::FiveTuple(five_tuple) => println!(
            "FiveTuple: {:?}",
            (
                five_tuple.source_ip.to_string(),
                five_tuple.destination_ip.to_string(),
                five_tuple.source_port,
                five_tuple.destination_port,
                five_tuple.protocol
            )
        ),
    }
}
fn main() {
    print_data(SourceData::IP("192.168.0.1".into()));
    print_data(SourceData::FiveTuple(FiveTuple {
        source_ip: "192.168.0.1".into(),
        destination_ip: "192.168.0.2".into(),
        source_port: 55555,
        destination_port: 53,
        protocol: Protocol::UDP,
    }));
    print_data(SourceData::FiveTuple(FiveTuple {
        source_ip: "127.0.0.1".into(),
        destination_ip: "127.0.0.1".into(),
        source_port: 54321,
        destination_port: 8080,
        protocol: Protocol::TCP,
    }));
}
